front/pages/[user]/a/[sId]/execute/index.jsx

```diff
@@ -0,0 +1,403 @@
+        <div className="flex flex-initial mt-2">
+          <MainTab
+            app={{ sId: app.sId, name: app.name }}
+            currentTab="Execute"
+            user={user}
+          />
+        </div>
+        <div className="w-full max-w-5xl mt-4 mx-auto">
+          <div className="flex flex-auto flex-col mx-2 sm:mx-4 lg:mx-8">
+            <ul className="space-y-2">
+              {inputDatasetKeys.map((k) => (
+                <li key={k} className="space-y-[1px]">
+                  <ExecuteInput
+                    inputName={k}
+                    inputValue={inputData[k]}
+                    onChange={(value) => handleValueChange(k, value)}
+                  />
+                </li>
+              ))}
+            </ul>
+            <br /> {/* TODO: no br */}
+            {executionLogs.blockOrder.length ? (
+              <ExecuteOutput
+                executionLogs={executionLogs}
+                expandedByBlockName={outputExpandedByBlockName}
+                onToggleExpand={(blockName) => {
+                  setOutputExpandedByBlockName((prev) => {
+                    const newExpanded = { ...prev };
+                    newExpanded[blockName] = !newExpanded[blockName];
+                    return newExpanded;
+                  });
+                }}
+              />
+            ) : null}
+            <div className="static inset-auto static inset-auto right-0 hidden sm:flex flex-initial items-center pr-2 sm:pr-0">
+              <div className="-mr-2 sm:mr-0">
+                <ActionButton
+                  disabled={isRunning || !isInputDataValid()}
+                  onClick={() => handleRun()}
+                >
+                  <PlayCircleIcon className="-ml-1 mr-1 h-5 w-5 mt-0.5" />
+                  Run
+                </ActionButton>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </AppLayout>
+  );
+}
+
+export async function getServerSideProps(context) {
+  const appRes = await fetch(
+    `${URL}/api/apps/${context.query.user}/${context.query.sId}`,
+    {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Cookie: context.req.headers.cookie,
+      },
+    }
+  );
+
+  if (appRes.status === 404) {
+    return {
+      notFound: true,
+    };
+  }
+
+  const app = await appRes.json();
+
+  const savedSpecification = JSON.parse(app.app.savedSpecification || "[]");
+  const inputBlock = savedSpecification.find((block) => block.type === "input");
+  const inputDatasetName = inputBlock.config.dataset;
+
+  const inputDatasetRes = await fetch(
+    `${URL}/api/apps/${context.query.user}/${context.query.sId}/datasets/${inputDatasetName}/latest`,
+    {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Cookie: context.req.headers.cookie,
+      },
+    }
+  );
+  const inputDataset = await inputDatasetRes.json();
+
+  const [specRes] = await Promise.all([
+    fetch(
```

Comment:
You can probably parallelise input and spec ?

Reply:
Indeed

Reply:
Actually now that `inputDatasetRes` is only computed if `inputBlock` isn't null, it's not that straightforward

Reply:
No longer relevant since I don't have to fetch spec. I can use `savedSpecification` from the `app`